tests/Unit/Evolution/Systems/ScalarWave/Test_Characteristics.cpp

```diff
@@ -0,0 +1,388 @@
+
+#include <array>
+#include <cstddef>
+#include <limits>
+#include <memory>
+#include <pup.h>
+
+#include "DataStructures/DataBox/Prefixes.hpp"  // IWYU pragma: keep
+#include "DataStructures/DataVector.hpp"        // IWYU pragma: keep
+#include "DataStructures/Tensor/EagerMath/Magnitude.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Domain/CoordinateMaps/Affine.hpp"
+#include "Domain/CoordinateMaps/CoordinateMap.hpp"
+#include "Domain/CoordinateMaps/ProductMaps.hpp"
+#include "Domain/LogicalCoordinates.hpp"
+#include "Domain/Mesh.hpp"
+#include "Evolution/Systems/ScalarWave/Characteristics.hpp"
+#include "Evolution/Systems/ScalarWave/Tags.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/WaveEquation/PlaneWave.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/WaveEquation/RegularSphericalWave.hpp"
+#include "PointwiseFunctions/MathFunctions/Gaussian.hpp"
+#include "PointwiseFunctions/MathFunctions/MathFunction.hpp"
+#include "Utilities/MakeWithValue.hpp"
+#include "Utilities/TaggedTuple.hpp"
+#include "tests/Unit/Pypp/CheckWithRandomValues.hpp"
+#include "tests/Unit/Pypp/SetupLocalPythonEnvironment.hpp"
+
+// IWYU pragma: no_forward_declare Tags::dt
+// IWYU pragma: no_forward_declare Tensor
+// IWYU pragma: no_forward_declare Variables
+
+namespace {
+template <size_t Index, size_t Dim>
+Scalar<DataVector> speed_with_index(
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& normal) {
+  return Scalar<DataVector>{
+      ScalarWave::CharacteristicSpeedsCompute<Dim>::function(normal)[Index]};
+}
+
+template <size_t Dim>
+void test_characteristic_speeds() noexcept {
+  const DataVector used_for_size(5);
+  pypp::check_with_random_values<1>(speed_with_index<0, Dim>, "Characteristics",
+                                    "char_speed_upsi", {{{-10.0, 10.0}}},
+                                    used_for_size);
+  pypp::check_with_random_values<1>(speed_with_index<1, Dim>, "Characteristics",
+                                    "char_speed_uzero", {{{-10.0, 10.0}}},
+                                    used_for_size);
+  pypp::check_with_random_values<1>(speed_with_index<3, Dim>, "Characteristics",
+                                    "char_speed_uminus", {{{-10.0, 10.0}}},
+                                    used_for_size);
+  pypp::check_with_random_values<1>(speed_with_index<2, Dim>, "Characteristics",
+                                    "char_speed_uplus", {{{-10.0, 10.0}}},
+                                    used_for_size);
+}
+
+// Test return-by-reference char speeds by comparing to analytic solution
+template <size_t Dim>
+void test_characteristic_speeds_analytic(
+    const size_t grid_size_each_dimension) noexcept {
+  // Setup mesh
+  Mesh<Dim> mesh{grid_size_each_dimension, Spectral::Basis::Legendre,
+                 Spectral::Quadrature::GaussLobatto};
+  const auto x_logical = logical_coordinates(mesh);
```

Comment:
I think you can delete this variable, it is only used in the line below ...

Reply:
done